lib/file-source/src/fingerprinter.rs

```diff
@@ -69,6 +70,74 @@ impl From<u64> for FileFingerprint {
     }
 }
 
+#[derive(Debug, Copy, Clone)]
+enum SupportedCompressionAlgorithms {
+    GZIP,
+}
+
+impl SupportedCompressionAlgorithms {
+    fn values() -> Vec<SupportedCompressionAlgorithms> {
+        vec![SupportedCompressionAlgorithms::GZIP]
+    }
+
+    fn magic_header_bytes(&self) -> &'static [u8] {
+        match self {
+            SupportedCompressionAlgorithms::GZIP => &[0x1f, 0x8b],
+        }
+    }
+}
+
+trait UncompressedReader {
+    fn check(fp: &mut File) -> Option<SupportedCompressionAlgorithms>;
+    fn reader<'a>(fp: &'a mut File) -> Box<dyn BufRead + 'a>;
+}
+
+struct UncompressedReaderImpl;
+impl UncompressedReader for UncompressedReaderImpl {
+    fn check(fp: &mut File) -> Option<SupportedCompressionAlgorithms> {
+        SupportedCompressionAlgorithms::values()
+            .iter()
+            .find_map(|compression_algorithm| {
+                // magic headers for algorithms can be of different lengths, and using a buffer too long could exceed the length of the file
+                // so instantiate and check the various sizes independently
+                let magic_header_bytes = compression_algorithm.magic_header_bytes();
+                let mut magic = vec![0u8; magic_header_bytes.len()];
+
+                if fp.read_exact(&mut magic).is_ok()
+                    && fp.seek(SeekFrom::Start(0)).is_ok()
```

Comment:
I think we always need to seek back even if the `read_exact` failed, no?

Also, if the seek fails, I think we should return an error (i.e. change the function signature here to return a `Result`) since the file pointer will be in an invalid state.

Reply:
Yes, that's a good point. Will change.